src/arkode/arkode_arkstep.c

```diff
@@ -3219,6 +3226,203 @@ int arkStep_ComputeSolutions_MassFixed(ARKodeMem ark_mem, sunrealtype* dsmPtr)
   return (ARK_SUCCESS);
 }
 
+/*---------------------------------------------------------------
+  Utility routines for interfacing with SUNStepper
+  ---------------------------------------------------------------*/
+
+int ARKStepCreateSUNStepper(void* inner_arkode_mem, SUNStepper* stepper)
+{
+  int retval;
+  ARKodeMem ark_mem;
+  ARKodeARKStepMem step_mem;
+
+  retval = arkStep_AccessStepMem(inner_arkode_mem, "ARKStepCreateSUNStepper",
+                                 &ark_mem, &step_mem);
+  if (retval)
+  {
+    arkProcessError(NULL, ARK_ILL_INPUT, __LINE__, __func__, __FILE__,
+                    "The ARKStep memory pointer is NULL");
+    return ARK_ILL_INPUT;
+  }
+
+  retval = SUNStepper_Create(ark_mem->sunctx, stepper);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  retval = SUNStepper_SetContent(*stepper, inner_arkode_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  retval = SUNStepper_SetAdvanceFn(*stepper, arkStep_SUNStepperAdvance);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  retval = SUNStepper_SetFullRhsFn(*stepper, arkStep_SUNStepperFullRhs);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  retval = SUNStepper_SetResetFn(*stepper, arkStep_SUNStepperReset);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  return (ARK_SUCCESS);
+}
+
+/*------------------------------------------------------------------------------
+  arkStep_SUNStepperAdvance
+
+  Implementation of SUNStepperAdvanceFn to advance the inner (fast)
+  ODE IVP.
+  ----------------------------------------------------------------------------*/
+
+int arkStep_SUNStepperAdvance(SUNStepper stepper, sunrealtype t0,
+                              sunrealtype tout, N_Vector y, sunrealtype* tret,
+                              int* stop_reason)
+{
+  void* arkode_mem;           /* arkode memory             */
+  sunrealtype tshift, tscale; /* time normalization values */
+  N_Vector* forcing;          /* forcing vectors           */
+  int nforcing;               /* number of forcing vectors */
+  int retval;                 /* return value              */
+
+  /* extract the ARKODE memory struct */
+  retval = SUNStepper_GetContent(stepper, &arkode_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* get the forcing data */
+  retval = SUNStepper_GetForcingData(stepper, &tshift, &tscale, &forcing,
+                                     &nforcing);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* set the inner forcing data */
+  retval = arkStep_SetInnerForcing(arkode_mem, tshift, tscale, forcing, nforcing);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* set the stop time */
+  retval = ARKStepSetStopTime(arkode_mem, tout);
```

Comment:
Is this needed? Maybe it could be an option. For operator splitting, it would be preferable not to have this.

Reply:
In our team discussions we defined the `Advance` function to operate in this way. That is, we said `Advance` would evolve to precisely `tout` as opposed to `Evolve` which would potentially step past `tout` and interpolate. `MRIStep` needs this.  It sounds like you need an `Evolve` function for the operator splitting stepper.

Reply:
Ah I didn't realize we would have both `Advance` and `Evolve`. I that case, this is not an issue.

Reply:
In the last discussion we dropped `Evolve` because it seemed like we did not need it. @Steven-Roberts for the splitting methods, would `Advance` still be sufficient or do you need to step past the time the method is supposed to integrate to?

Reply:
`Advance` alone would be sufficient, but then I don't see a way to handle symmetric operator splitting methods most efficiently. As an example of the functionality needed, consider integration a stepper from `t=0.5` to `t=1.5` with output needed at `t=1`. Two calls to `Evolve` would be ideal for this, while `Advance` might add more function evaluations by forcing the first call to end exactly at `t=1`. Originally, I thought adding an interpolation function to the stepper was the way to fix this, but on second thought I don't think it's reasonable to have interpolation data across the entire tspan passed to `Advance`.

Reply:
That's another way to look at it. Multiple outputs with one `Advance` call would work, but the function signature would start to look very different than `ARKStepEvolve` in normal mode. An alternative is to add a `SetTStop` function to `SUNStepper`.

Reply:
> Adding a stop time would also work but could lead to extra RHS evaluations.

I could only see this leading to extra RHS evaluations if it was set unnecessarily.

Reply:
If going from `t = 0.5` to `t = 1.5` with an output at `t = 1` and using t-stop at `t = 1`, the integrator will need to adjust its step to stop exactly at `t = 1` and so will need an extra step compared to stepping over and interpolating the output at `t = 1`.

Reply:
I was thinking that situation would be handled like this
```
SetTStop(stepper, 1.5);
Advance(stepper, 0.5, 1, , y, ...);
// The inner solve may have overstepped t=1, but it returns interpolated solution at t=1
// Do something with y
Advance(stepper, 0.5, 1, , y, ...);
// This will end exactly at t=1.5 and not overstep
```

Reply:
Ah, I see what you mean now. Yes, that would work as well.

Reply:
That is precisely how `ARKStepEvolve` or any other of the other package "evolve"-type functions work.

Reply:
Right, it's sounding like we need `Evolve` + `SetStopTime` rather than `Advance`

Reply:
Originally we talked about having four functions:

- `Evolve`:  step to tout, possibly passing it and returning interpolated solution at t=tout 
- `Advance`: step to tout, stopping exactly at tout and return
- `Step`: take one _successful_ step (or fail trying to) and return
- `TryStep`: take exactly one internal step and return

We said we did not have a use for `Evolve` yet, so I removed it from the `SUNStepper` API for now. However, it seems the operator splitting stepper does necessitate an `Evolve` function PLUS a `SUNStepper_SetTStop` function. 

If we add a `SUNStepper_SetTStop` function, then it should only be used with `Evolve`, NOT `Advance` otherwise there is no reason to have `Advance` as a function since it would work just like `Evolve`.

Reply:
I'm leaning toward `Evolve` + `SetStopTime` since `Advance` (or even `Evolve`) with multiple outputs could just be a wrapper looping over the output times and calling the one output `Evolve` with the right output vector.

Reply:
> However, it seems the operator splitting stepper does necessitate an Evolve function PLUS a SUNStepper_SetTStop function.

It does need the `Evolve` function. `SetTStop` would only be needed if `Advance` was removed (as I could be achieved by combining `Evolve` and `SetTStop`).

So I think the question is which set of functions to use
1. `Evolve`, 'Advance`, `Step`, `TryStep` or
2. `Evolve`, `SetTStop`, `Step`, `TryStep`

Reply:
I'm a bit late to the party, but it sounds to me like `Evolve` + `SetStopTime` is the way to go.  I'll note that like Steven, in the MERK multirate time-stepping methods I evolve over a longer duration, periodically storing the solution for some intermediate stages.  It would be perfectly handled by this approach, whereas right now I anticipate that the MRIStepInnerStepper may take a small number of extra steps as it stops to return intermediate solutions.  I'll defer to the rest of the group on the need for `Step` and `TryStep` usage modes.